src/main/java/net/logstash/logback/appender/AsyncDisruptorAppender.java

```diff
@@ -422,57 +450,127 @@ public void stop() {
         }
 
-        if (!this.disruptor.getRingBuffer().tryPublishEvent(this.eventTranslator, event)) {
-            long consecutiveDropped = this.consecutiveDroppedCount.incrementAndGet();
-            if ((consecutiveDropped) % this.droppedWarnFrequency == 1) {
-                addWarn("Dropped " + consecutiveDropped + " events (and counting...) due to ring buffer at max capacity [" + this.ringBufferSize + "]");
-            }
-            fireEventAppendFailed(event, RING_BUFFER_FULL_EXCEPTION);
-        } else {
-            long endTime = System.nanoTime();
+        if (enqueueEvent(event)) {
+            // Enqueue success - notify if we had errors previously
+            //
             long consecutiveDropped = this.consecutiveDroppedCount.get();
             if (consecutiveDropped != 0 && this.consecutiveDroppedCount.compareAndSet(consecutiveDropped, 0L)) {
                 addWarn("Dropped " + consecutiveDropped + " total events due to ring buffer at max capacity [" + this.ringBufferSize + "]");
             }
-            fireEventAppended(event, endTime - startTime);
+            
+            // Notify parties
+            //
+            fireEventAppended(event, System.nanoTime() - startTime);
+            
+        } else {
+            // Log a warning status about the failure
+            //
+            long consecutiveDropped = this.consecutiveDroppedCount.incrementAndGet();
+            if ((consecutiveDropped % this.droppedWarnFrequency) == 1) {
+                addWarn("Dropped " + consecutiveDropped + " events (and counting...) due to ring buffer at max capacity [" + this.ringBufferSize + "]");
+            }
+            
+            // Notify parties
+            //
+            fireEventAppendFailed(event, RING_BUFFER_FULL_EXCEPTION);
         }
     }
 
     protected void prepareForDeferredProcessing(Event event) {
         event.prepareForDeferredProcessing();
     }
 
+    /**
+     * Enqueue the given {@code event} in the ring buffer.
+     * 
+     * @param event the {@link Event} to enqueue
+     * @return {@code true} when the even is successfully enqueued in the ring buffer
+     */
+    protected boolean enqueueEvent(Event event) {
+        long timeout = this.appendTimeout.getMilliseconds() < 0 ? Long.MAX_VALUE : System.currentTimeMillis() + this.appendTimeout.getMilliseconds();
+        
+        while (isStarted() && !this.disruptor.getRingBuffer().tryPublishEvent(this.eventTranslator, event)) {
+            // Check for timeout
+            //
+            if (System.currentTimeMillis() >= timeout) {
+                return false;
+            }
+            
+            // Wait before retry
+            //
+            long waitDuration = Math.min(this.appendRetryFrequency.getMilliseconds(), System.currentTimeMillis() - timeout);
+            if (waitDuration > 0) {
+                LockSupport.parkNanos(waitDuration * 1_000_000L);
+            }
+        }
+        
+        return true;
+    }
```

Comment:
This method could return true (incorrectly) if the appender is stopped while the loop is processing.

Also, I think the subtraction order is reversed in the waitDuration calculation on line 566.  I think it should be `timeout - System.currentTimeMillis()` since at this point currentTimeMillis < timeout (due to the check on line 560).  And, after correcting the order, the additional timeout check can be replaced by an else block on the waitDuration check.

So, addressing those comments, here's a possibility:

```java
    protected boolean enqueueEvent(Event event) {
        long deadline = this.appendTimeout.getMilliseconds() < 0 ? Long.MAX_VALUE : System.currentTimeMillis() + this.appendTimeout.getMilliseconds();

        while (isStarted()) {
            if (this.disruptor.getRingBuffer().tryPublishEvent(this.eventTranslator, event)) {
                return true;
            }
            // Wait before retry
            //
            long waitDuration = Math.min(this.appendRetryFrequency.getMilliseconds(), deadline - System.currentTimeMillis());
            if (waitDuration > 0) {
                LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(waitDuration));
            } else {
                break;
            }
        }

        return false;
    }
```

Reply:
> This method could return true (incorrectly) if the appender is stopped while the loop is processing.

Correct. And this was on purpose... 

First of all, events emitted by the application after the appender is stopped are silently discarded (a first warn is emitted tho). Loosing an event because the retry process is aborted during shutdown looks the same to me... so I was ok keeping it silent.
Now suppose the method return false when the appender is stopped as you propose, a warn status will be emitted saying an event is _dropped because the buffer is full_ (which is correct). I didn't feel comfortable with this and initially thought I had to return the actual cause to provide a more accurate cause... but that made the thing more complex. 

With al that said, I now realise that a _buffer full_ message is enough and we don't probably need anything more precise (what would be the added value?).

> I think the subtraction order is reversed in the waitDuration calculation on line 566
Correct... this shows that my test cases are a bit too optimistic :-(

Reply:
Ok.  Thanks for the explanation.

I commented on this because the documented behavior...

> @return {@code true} when the even is successfully enqueued in the ring buffer

... is different than the implementation.   If the appender is shutdown while attempting to enqueue, the event won't be enqueued, but the method still returns true (contrary to the documented behavior).  Therefore, this seemed like a bug.

Also, when looking at the method in isolation (without looking at the calling code), returning true when the event is not enqueued seems... odd.   So the documentation seemed correct to me, and the implementation seemed incorrect.

Perhaps if the method detects a shutdown, the method could throw an exception, rather than returning true?  Then the caller could handle it by swallowing it (and not logging an event dropped warning), or logging a different warning.

I'm ok with whatever you decide, but regardless of your decision, please update the documentation to describe all cases.

Reply:
I moved back the "enqueue" logic inside the "append()" method as it was before - the method looks as it was before with an additional while loop.
Regarding the appender stopped when coming back from sleep, I decided to simply "warn" the user with the same message as what would be emitted if the application attempts to log juste after the appender is stopped.